File: reference-solutions/python/janelas-de-manutencao.py
import sys


def solve(data: str) -> str:
    values = list(map(int, data.split()))
    n = values[0]
    if n == 0:
        return "0 0"
    intervals = sorted((values[2 * i + 1], values[2 * i + 2]) for i in range(n))
    start, end = intervals[0]
    total = longest = 0
    for next_start, next_end in intervals[1:]:
        if next_start <= end:
            end = max(end, next_end)
        else:
            total += end - start
            longest = max(longest, end - start)
            start, end = next_start, next_end
    total += end - start
    longest = max(longest, end - start)
    return f"{total} {longest}"


if __name__ == "__main__":
    print(solve(sys.stdin.read()))

File: reference-solutions/typescript/janelas-de-manutencao.ts
import { readFileSync } from "node:fs";

export function solve(input: string): string {
  const values = input.trim().split(/\s+/).map(Number);
  const n = values[0]!;
  if (n === 0) return "0 0";
  const intervals: Array<[number, number]> = [];
  for (let i = 0; i < n; i += 1) intervals.push([values[2 * i + 1]!, values[2 * i + 2]!]);
  intervals.sort((a, b) => a[0] - b[0] || a[1] - b[1]);

  let [start, end] = intervals[0]!;
  let total = 0;
  let longest = 0;
  for (let i = 1; i < n; i += 1) {
    const [nextStart, nextEnd] = intervals[i]!;
    if (nextStart <= end) {
      // Inclui janelas contíguas; intervalos contidos nunca reduzem o fim.
      end = Math.max(end, nextEnd);
    } else {
      total += end - start;
      longest = Math.max(longest, end - start);
      [start, end] = [nextStart, nextEnd];
    }
  }
  total += end - start;
  longest = Math.max(longest, end - start);
  return `${total} ${longest}`;
}

console.log(solve(readFileSync(0, "utf8")));
